_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+        //Filter it down to the current record
+        .Where(i => i.FasterOfferWithLinq.Oid == Oid)
+        //Sum the total hours
+        .Sum(m => m.Hours);
+
+    [Association, Aggregated]
+    public XPCollection<FasterOfferItemWithLinq> OfferItems => GetCollection<FasterOfferItemWithLinq>(nameof(OfferItems));
+}
+
+public class FasterOfferItemWithLinq : BaseObject
+{
+    public FasterOfferItemWithLinq(Session session) : base(session) { }
+
+    private int _Hours;
+    [Persistent("Hours")]
+    public int Hours { get => _Hours; set => SetPropertyValue(nameof(Hours), ref _Hours, value); }
+
+    private FasterOfferWithLinq _FasterOfferWithLinq;
+    [Persistent("FasterOfferWithLinq"), Association]
+    public FasterOfferWithLinq FasterOfferWithLinq { get => _FasterOfferWithLinq; set => SetPropertyValue(nameof(FasterOfferWithLinq), ref _FasterOfferWithLinq, value); }
+}
+```
+
+This will result in much leaner queries. Still does not solve the N+1 problem though:
+
+```txt
+19.09.20 10:58:05.069 Executing sql 'select N0."Oid",N0."Name",N0."OptimisticLockField",N0."GCRecord" from "dbo"."FasterOfferWithLinq" N0 where N0."GCRecord" is null'
+19.09.20 10:58:05.070 Result: rowcount = 300, total = 11992, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4792, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
+19.09.20 10:58:05.071 Executing sql 'select top 1 sum(N0."Hours") from "dbo"."FasterOfferItemWithLinq" N0 where (N0."GCRecord" is null and (N0."FasterOfferWithLinq" = @p0))' with parameters {a3797700-1fcf-429f-9c34-005322afa892}
+19.09.20 10:58:05.075 Result: rowcount = 1, total = 4, SubQuery(Sum,N0.{Hours,Int32},) = 4
+19.09.20 10:58:05.076 Executing sql 'select top 1 sum(N0."Hours") from "dbo"."FasterOfferItemWithLinq" N0 where (N0."GCRecord" is null and (N0."FasterOfferWithLinq" = @p0))' with parameters {3683bfc2-7314-452f-be51-00692f123965}
+.
+.
+.
+19.09.20 10:58:06.497 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithLinq" N0 where N0."GCRecord" is null'
+19.09.20 10:58:06.497 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 10:58:06.498 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItemWithLinq" N0 where N0."GCRecord" is null'
+19.09.20 10:58:06.512 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 10:58:07.919 Executing sql 'select top 1 sum(N0."Hours") from "dbo"."FasterOfferItemWithLinq" N0 where (N0."GCRecord" is null and (N0."FasterOfferWithLinq" = @p0))' with parameters {820161dc-8568-412c-bf84-c4f148e8122f}
+19.09.20 10:58:07.924 Result: rowcount = 1, total = 4, SubQuery(Sum,N0.{Hours,Int32},) = 4
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.156 seconds
+2. Sort by `HourSum`: 1.554 seconds
+
+As you can see, it performs a lot better. And for the *most parts* of your application this will be totally fine (esp. if you arn't dealing with too much records, or lists that arn't used that frequently).
```

Comment:
```suggestion
As you can see, it performs a lot better. And for the *most parts* of your application, this will be totally fine (esp. if you aren't dealing with too many records, or lists that aren't used that frequently).
```